File: jsrc/machine/SxxMachine/KPTerm.java
package SxxMachine;

/**
 * Top element of the Prolog term hierarchy. Describes a simple or compound ter
 * like: X,a,13,f(X,s(X)),[a,s(X),b,c], a:-b,c(X,X),d, etc.
 */
public abstract class KPTerm extends TermData implements Undoable {

	public void oopsy() {
		new Exception().printStackTrace();
	}
	public static void soopsy() {
		new Exception().printStackTrace();
	}

	abstract // @Override
	public Term toClone() throws CloneNotSupportedException;
	
	@Override
	public int hashCode() {
		return termHashCode();		
	}
	
	public boolean equals(Object o) {
		return o==this;
	}
	
	@Override
	public String toString() {
		return pprint();
	}
	public abstract int termHashCode();


	/**
	 * returns or fakes an arity for all subtypes
	 */
	abstract public int arityOrType();

	/**
	 * Dereferences if necessary. If multi-threaded, this should be synchronized
	 * otherwise vicious non-reentrancy problems may occur in the presence of GC and
	 * heavy multi-threading!!!
	 */
	public Term dref() { // synchronized !!!
		return (Term) this;
	}

	public abstract boolean bind(Term that, KPTrail trail);

	/** Unify dereferenced */
	public abstract boolean Unify_TO(Term that, KPTrail trail);

	/** Dereference and unify_to */
	public final boolean DO_Unify(Term that, KPTrail trail) {
	  if(that==this) return true;
	  Term thisd = dref();
	  Term thatd = that.dref();
	  if(thatd==thisd) return true;
		return thisd.Unify_TO(thatd, trail);
	}

	public void undo() { // does nothing
	}

	// public abstract boolean eq(Term that);

	public Term carTokenOrSelf() {
		return (Term) this;
	}

	Term toTerm() {
		return (Term) this;
	}

	public Clause toClause() {
		return new Clause((Term) this, Prolog.True);
	}

	public boolean isClause() {
		return false;
	}

	public static Term fromString(String s) {
		return Clause.clauseFromString(s).toTerm();
	}

	/**
	 * Tests if this term unifies with that. Bindings are trailed and undone after
	 * the test. This should be used with the shared term as this and the new term
	 * as that. Synchronization makes sure that side effects on the shared term are
	 * not interfering, i.e as in: SHARED.matches(NONSHARED,trail).
	 * 
	 */
	// synchronized
	public boolean matches(Term that) {
		return matches(that, new KPTrail());
	}

	public boolean matches(Term that, KPTrail trail) {
		int oldtop = trail.size();
		boolean ok = DO_Unify(that, trail);
		trail.unwind(oldtop);
		return ok;
	}

	/**
	 * Returns a copy of the result if the unification of this and that. Side
	 * effects on this and that are undone using trailing of bindings..
	 * Synchronization happens over this, not over that. Make sure it is used as
	 * SHARED.matching_copy(NONSHARED,trail).
	 */
	// synchronized

	public Term matching_copy(Term that) {
		KPTrail trail = new KPTrail();
		boolean ok = DO_Unify(that, trail);
		// if(ok) that=that.copy();
		if (ok)
			that = copy();
		trail.unwind(0);
		return (ok) ? that : null;
	}

	/**
	 * Defines the reaction to an agent recursing over the structure of a term.
	 * <b>This</b> is passed to the agent and the result of the action is returned.
	 * Through overriding, for instance, a Fun term will provide the recursion over
	 * its arguments, by applying the action to each of them.
	 * 
	 * @see StructureTerm
	 */
	public Term reaction(Term agent) {
		Term T = agent.action((Term) this);
		return T;
	}

	/**
	 * Identity action.
	 */
	public Term action(Term that) {
		return that;
	}

	/**
	 * Returns a copy of a term with variables standardized apart (`fresh
	 * variables').
	 */
	// synchronized
	public Term copy() {
		return reaction(new Copier());
	}

	/**
	 * Returns '[]'(V1,V2,..Vn) where Vi is a variable occuring in this Term
	 */
	public Term varsOf() {
		return (new Copier()).getMyVars((Term) this);
	}

	/**
	 * Replaces variables with uppercase constants named `V1', 'V2', etc. to be read
	 * back as variables.
	 */
	public Term numbervars() {
		return copy().reaction((Term) new VarNumberer());
	}

	/**
	 * Prints out a term to a String with variables named in order V1, V2,....
	 */
	public String toUnquoted() {
		Term t = numbervars();
		if(t==this) return pprint();
		return t.pprint();
	}

	/*
	 * Returns an unquoted version of toString()
	 */
	abstract public String pprint();

	/**
	 * Returns a string key used based on the string name of the term. Note that the
	 * key for a clause AL-B,C. is the key insted of ':-'.
	 */
	public String getKey() {
		return toUnquoted();
	}

	/**
	 * Java Object wrapper. In particular, it is used to wrap a Thread to hide it
	 * inside a Prolog data object.
	 */
	public Object toObject() {
		return dref();
	}

	/*
	 * Just to catch the frequent error when the arg is forgotten while definig a
	 * builtin. Being final, it will generate a compile time error if this happens
	 */
	final int exec() {

		return -1;
	}

	/**
	 * Executed when a builtin is called. Needs to be overriden. Returns a run-time
	 * warning if this is forgotten.
	 */

	public int exec(Prog p) {
		// IO.println("this should be overriden, prog="+p);
		return -1;
	}

	static final Nonvar stringToChars(String s) {
		if (0 == s.length())
			return Prolog.Nil;
		ListTerm l = CONS(TermData.Long((s.charAt(0))), Prolog.Nil);
		Term curr =  l;
		for (int i = 1; i < s.length(); i++) {
			StructureTerm tail = StructureTerm.createCons(Long((s.charAt(i))), Prolog.Nil);
			curr.setArg(1, tail);
			curr = tail;
		}
		return l;
	}

	public Nonvar toChars() {
		return stringToChars(toUnquoted());
	}

	/**
	 * Converts a list of character codes to a String.
	 */
	public static String charsToString(Nonvar Cs) {
		StringBuffer s = new StringBuffer("");

		while (!(Cs.isNil())) {
			if (!(Cs.isCons()))
				return null;
			Nonvar head = (Nonvar) Expect.asCons(Cs).ArgDeRef(0);
			
			if (!(head.isNumber()))
				return null;
			char c = (char) Expect.asInt(head).longValue();
			s.append(c);
			Cs = (Nonvar) Expect.asCons(Cs).ArgDeRef(1);
		}

		return s.toString();
	}

	public boolean isCons() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isNil() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isBuiltin() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isConj() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isLong() {
		// TODO Auto-generated method stub
		return ((Term)this).isInteger();
	}

	public int intValue() {
		// TODO Auto-generated method stub
		return 0;
	}

	public boolean isConst() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isStructure() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isTrueProc() {
		return false;
	}

	public boolean isNumber() {
		// TODO Auto-generated method stub
		return false;
	}

	abstract public String name();

	public boolean isFloat() {
		return false;
	}

	public boolean isSystemObject() {
		return false;
	}

	public boolean isFluent() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isObject() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isVar() {
		// TODO Auto-generated method stub
		return false;
	}

	public boolean isCharReader() {
		// TODO Auto-generated method stub
		return false;
	}

	public Term toValue() {
		// TODO Auto-generated method stub
		return (Term) this;
	}

	public double doubleValue() {
		// TODO Auto-generated method stub
		return 0; 
	}

	abstract public boolean equalsTerm(Term aneof) ;

	final public SymbolTerm asConst() {
		// TODO Auto-generated method stub
		return (SymbolTerm)this;
	}

	public Var toVar() {
		// TODO Auto-generated method stub
		return (Var)this;
	}

	public Term ArgDeRef(int i) {
		return Expect.asStruct((Term)this).ArgDeRef(i);
	}

	public boolean isFunctor(String string) {
		// TODO Auto-generated method stub
		return name().equals(string);
	}
	public Source asSource() {
		// TODO Auto-generated method stub
		return (Source)this;
	}
}

File: src/machine/SxxMachine/KPTerm.cpp
using namespace std;

#include "KPTerm.h"
#include "Term.h"
#include "../../kernelprolog/main/SxxMachine/KPTrail.h"
#include "../../kernelprolog/terms/SxxMachine/Clause.h"
#include "Prolog.h"
#include "../../kernelprolog/terms/SxxMachine/Copier.h"
#include "../../kernelprolog/main/SxxMachine/Builtins.h"
#include "../../kernelprolog/terms/SxxMachine/VarNumberer.h"
#include "../../kernelprolog/terms/SxxMachine/Prog.h"
#include "Nonvar.h"
#include "ListTerm.h"
#include "StructureTerm.h"
#include "../../kernelprolog/terms/SxxMachine/Expect.h"
#include "SymbolTerm.h"
#include "../../kernelprolog/terms/SxxMachine/Var.h"
#include "../../kernelprolog/terms/SxxMachine/Source.h"
#include "StringBuilder.h"

namespace SxxMachine
{

	void KPTerm::oopsy()
	{
//JAVA TO C++ CONVERTER TODO TASK: This exception's constructor requires an argument:
//ORIGINAL LINE: new Exception().printStackTrace();
		(runtime_error())->printStackTrace();
	}

	void KPTerm::soopsy()
	{
//JAVA TO C++ CONVERTER TODO TASK: This exception's constructor requires an argument:
//ORIGINAL LINE: new Exception().printStackTrace();
		(runtime_error())->printStackTrace();
	}

	int KPTerm::hashCode()
	{
		return termHashCode();
	}

	bool KPTerm::equals(any o)
	{
		return o == this;
	}

	wstring KPTerm::toString()
	{
		return pprint();
	}

	Term* KPTerm::dref()
	{ // synchronized !!!
		return static_cast<Term*>(this);
	}

	bool KPTerm::DO_Unify(Term* that, KPTrail* trail)
	{
		if(that == this)
		{
			return true;
		}
		Term* thisd = dref();
		Term* thatd = that->dref();
		if(thatd == thisd)
		{
			return true;
		}
		return thisd->Unify_TO(thatd, trail);
	}

	void KPTerm::undo()
	{ // does nothing
	}

	Term* KPTerm::carTokenOrSelf()
	{
		return static_cast<Term*>(this);
	}

	Term* KPTerm::toTerm()
	{
		return static_cast<Term*>(this);
	}

	Clause* KPTerm::toClause()
	{
		return new Clause(static_cast<Term*>(this), Prolog::True);
	}

	bool KPTerm::isClause()
	{
		return false;
	}

	Term* KPTerm::fromString(const wstring& s)
	{
		return Clause::clauseFromString(s)->toTerm();
	}

	bool KPTerm::matches(Term* that)
	{
		KPTrail tempVar();
		return matches(that, &tempVar);
	}

	bool KPTerm::matches(Term* that, KPTrail* trail)
	{
		int oldtop = trail->size();
		bool ok = DO_Unify(that, trail);
		trail->unwind(oldtop);
		return ok;
	}

	Term* KPTerm::matching_copy(Term* that)
	{
		KPTrail* trail = new KPTrail();
		bool ok = DO_Unify(that, trail);
		// if(ok) that=that.copy();
		if(ok)
		{
			that = copy();
		}
		trail->unwind(0);
		return (ok) ? that : nullptr;
	}

	Term* KPTerm::reaction(Term* agent)
	{
		Term* T = agent->action(static_cast<Term*>(this));
		return T;
	}

	Term* KPTerm::action(Term* that)
	{
		return that;
	}

	Term* KPTerm::copy()
	{
		Copier tempVar();
		return reaction(&tempVar);
	}

	Term* KPTerm::varsOf()
	{
		Copier tempVar();
		return (&tempVar)->getMyVars(static_cast<Term*>(this));
	}

	Term* KPTerm::numbervars()
	{
		VarNumberer tempVar();
		return copy()->reaction(static_cast<Term*>(&tempVar));
	}

	wstring KPTerm::toUnquoted()
	{
		Term* t = numbervars();
		if(t == this)
		{
			return pprint();
		}
		return t->pprint();
	}

	wstring KPTerm::getKey()
	{
		return toUnquoted();
	}

	any KPTerm::toObject()
	{
		return dref();
	}

	int KPTerm::exec()
	{

		return -1;
	}

	int KPTerm::exec(Prog* p)
	{
		// IO.println("this should be overriden, prog="+p);
		return -1;
	}

	Nonvar* KPTerm::stringToChars(const wstring& s)
	{
		if(0 == s.length())
		{
			return Prolog::Nil;
		}
		ListTerm* l = CONS(TermData::Long((s[0])), Prolog::Nil);
		Term* curr = l;
		for(int i = 1; i < s.length(); i++)
		{
			StructureTerm* tail = StructureTerm::createCons(Long((s[i])), Prolog::Nil);
			curr->setArg(1, tail);
			curr = tail;
		}
		return l;
	}

	Nonvar* KPTerm::toChars()
	{
		return stringToChars(toUnquoted());
	}

	wstring KPTerm::charsToString(Nonvar* Cs)
	{
		StringBuilder* s = new StringBuilder("");

		while(!(Cs->isNil()))
		{
			if(!(Cs->isCons()))
			{
				return "";
			}
			Nonvar* head = static_cast<Nonvar*>(Expect::asCons(Cs)->ArgDeRef(0));

			if(!(head->isNumber()))
			{
				return "";
			}
			char c = static_cast<char>(Expect::asInt(head)->longValue());
			s->append(c);
			Cs = static_cast<Nonvar*>(Expect::asCons(Cs)->ArgDeRef(1));
		}

		return s->toString();
	}

	bool KPTerm::isCons()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isNil()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isBuiltin()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isConj()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isLong()
	{
		// TODO Auto-generated method stub
		return (static_cast<Term*>(this))->isInteger();
	}

	int KPTerm::intValue()
	{
		// TODO Auto-generated method stub
		return 0;
	}

	bool KPTerm::isConst()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isStructure()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isTrueProc()
	{
		return false;
	}

	bool KPTerm::isNumber()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isFloat()
	{
		return false;
	}

	bool KPTerm::isSystemObject()
	{
		return false;
	}

	bool KPTerm::isFluent()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isObject()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isVar()
	{
		// TODO Auto-generated method stub
		return false;
	}

	bool KPTerm::isCharReader()
	{
		// TODO Auto-generated method stub
		return false;
	}

	Term* KPTerm::toValue()
	{
		// TODO Auto-generated method stub
		return static_cast<Term*>(this);
	}

	double KPTerm::doubleValue()
	{
		// TODO Auto-generated method stub
		return 0;
	}

	SymbolTerm* KPTerm::asConst()
	{
		// TODO Auto-generated method stub
		return static_cast<SymbolTerm*>(this);
	}

	Var* KPTerm::toVar()
	{
		// TODO Auto-generated method stub
		return static_cast<Var*>(this);
	}

	Term* KPTerm::ArgDeRef(const int& i)
	{
		return Expect::asStruct(static_cast<Term*>(this))->ArgDeRef(i);
	}

	bool KPTerm::isFunctor(const wstring& string)
	{
		// TODO Auto-generated method stub
		return name() == string;
	}

	Source* KPTerm::asSource()
	{
		// TODO Auto-generated method stub
		return static_cast<Source*>(this);
	}
}
